ux-redesign: VM Details > Disks Card

sjd78 opened the pull request:
Closes: #775
~~Fixes: #851~~ (PR #863 fixed Issue #851), Fixes: #852

**What was added?**
In View:
 - Handle "no disks" scenario
 - `DiskListItem` renedered per disk attached to the VM.  It includes:
   - Disk state icon (active, inactive, locked)
   - Alias/Name
   - bootable label if the disk is bootable
   - provisioned size in GiB
   - Edit mode can be activated with the pencil icon based on the user's authorization to edit the VM

In Edit:
 - Create New Disk link available
 - Each existing disk is given an edit and delete icon
   - A disk may be deleted if the VM is down, it isn't in process of being deleted and it isn't locked
   - A disk may be edited as long as it isn't in process of being deleted and it isn't locked

Create Disk:
 - To create a disk, a new modal dialog is opened
 - Creation requires:
   - Disk alias (one will be suggested based on VM name and existing disks)
   - Disk size (defaults to 1GiB)
   - Storage domain (if the VM already has attached Disks, the most commonly used storage domain will be selected by default)
   - Disk format (raw && !sparse or cow && sparse)

Edit Disk:
 - To edit a disk, a new modal dialog is opened
 - The current size of the disk is always displayed
 - Edits are only allowed on:
   - Disk alias
   - If the disk is an **image** type, Disk size (only to extend/expand the disk image)
 - Once change to a disk have been submitted (asynchronously) the card is kept up to date with the disk's state (it'll show as locked while the disk image is being expanded)

Delete Disk:
 - The VM must be stopped/down to allow deletion of a VM disk
 - A delete confirmation modal verifies the action
 - While the deletion is in progress, the locked state of the disk is kept current on the card itself

**Legacy Disk Updates**
 - Adopted the redux/ovirtapi updates
 - Fixed the way disks are created so they are attached properly.  See [BZ1639630](https://bugzilla.redhat.com/show_bug.cgi?id=1639630).

**ovirtapi updates**
 - In transforms, created new helper function `cleanUndefined` to remove keys from Objects whose values are `undefined`.  It makes it a bit easier to work with ImmutableJS and doing `mergeDeepIn` operations of transformed objects.
 - All redux create and delete related things moved to the top level
 - Added the ability to EDIT/UPDATE a disk attachment/disk (new actions, sagas, VMS reducers, and API calls)

**NOTE:** webadmin allows changing a disk's Storage Domain, but under the "Move" operation.  We wont support that in webadmin so Storage Domain cannot be changed by a user after creation

bond95 commented on src/saga/disks.js:
```diff
@@ -0,0 +1,101 @@
+    const apiDisk = yield callExternalAction('disk', Api.disk, { payload: { diskId: disk.id } })
+    const edited = Api.Transforms.DiskAttachment.toInternal({ disk: apiDisk })
+
+    console.info('status', edited.status, 'edited disk to check', edited)
```
Remove, please.

bond95 commented on src/ovirtapi/index.js:
```diff
@@ -38,6 +39,8 @@ const OvirtApi = {
     return Transforms.VM.toInternal({ vm, includeSubResources: getSubResources })
   },
 
+  Transforms: Transforms,
```
I found. Use, please, `diskToInternal`, for consistency.

sjd78 replied:
I'd like to replace all the transform functions in the API with use of the actual `Transforms` object instead.  For now I suppose it isn't really necessary.

bond95 commented on src/saga/disks.js:
```diff
@@ -0,0 +1,101 @@
+import { CREATE_DISK_FOR_VM, REMOVE_DISK, EDIT_VM_DISK } from '../constants'
+import Api from 'ovirtapi'
+import { callExternalAction, delay } from './utils'
+import { fetchDisks } from '../sagas'
```
I think better to move `fetchDisks` function to this file.

sjd78 replied:
I'm not going to start moving saga functions around at this time.  This **disks.js** is really just the sagas file from the legacy disk components moved here so I kept it distinct.

Sometime in the future we can refactor the sagas and split them up nicely.

sjd78 commented on src/components/VmDetails/cards/DisksCard/DiskImageEditor.js:
```diff
@@ -0,0 +1,408 @@
+
+  changeSize (e) {
+    if (e.target.value < 0) {
+      e.target.value = this.state.size
```
It just prevents the user from setting a value less then 0.

Any ideas on how better to validate input?

bond95 replied:
Yes, for example:

```JavaScript
  changeSize (e) {
        let value = parseInt(e.target.value);
        value = value < 0 || isNaN(value) ? 0 : value;
        this.setState({ size: value })
        this.changesMade = true
    }
```

sjd78 replied:
While that will avoid setting the component state's size below 0, it doesn't prevent the display from going below 0.  I'll probably need to switch to a controlled `FormControl` to do things in a more straightforward manner.

sjd78 replied:
Switched `FormControl` from `defaultValue` to `value` (making the field controlled) so that is only displayed what is held in the `DiskImageEditor`'s state.  Makes the `changeSize` function very simple.

sjd78 commented on src/components/VmDetails/cards/DisksCard/index.js:
```diff
@@ -2,81 +2,260 @@ import React from 'react'
+        const isBeingDeleted = isDiskBeingDeleted(disk.get('id'), pendingTasks)
+        const isLocked = disk.get('status') === 'locked'
+
+        return {
```
I'm a bit confused.  Are you saying to put the VO in `DiskListItem`?  This is the same style I used in `NicsCard`.

sjd78 replied:
I moved it and it makes both components a bit better encapsulated.

sjd78 commented:
@bond95 - Please review my most recent changes.  Thanks!

sjd78 commented:
VM with a single bootable disk (in context on the VM Details dashboard page):
![screenshot-localhost-3000-2018 10 31-13-56-38](https://user-images.githubusercontent.com/3985964/47808515-ea03c480-dd14-11e8-8e51-dde21747a7fe.png)

When no disks are attached to a vm:
![screenshot-localhost-3000-2018 10 31-13-55-11 no disks](https://user-images.githubusercontent.com/3985964/47808554-f7b94a00-dd14-11e8-9ac5-975656241b75.png)

Disks card in edit mode (VM is down so delete is allowed):
![screenshot-localhost-3000-2018 10 31-13-58-25](https://user-images.githubusercontent.com/3985964/47808589-0bfd4700-dd15-11e8-866f-5f87eb90bf11.png)

Create New Disk modal with suggested name ("Charlie_Disk1" for VM named "Charlie") and storage domain:
![screenshot-localhost-3000-2018 10 31-13-59-14](https://user-images.githubusercontent.com/3985964/47808678-40710300-dd15-11e8-89b6-37ddf72b0154.png)

Editing a normal image disk (alias can be changed and disk can be expanded):
![screenshot-localhost-3000-2018 10 31-15-12-59](https://user-images.githubusercontent.com/3985964/47812692-76b38000-dd1f-11e8-98c2-0c37f3c4c148.png)

Editing a DirectLUN disk (only the alias can be changed):
![screenshot-localhost-3000-2018 10 31-14-00-17](https://user-images.githubusercontent.com/3985964/47808709-5979b400-dd15-11e8-9d87-948ecdb64f02.png)

Immediately after the disk is created but before it is ready to use (disk is locked):
OR Immediately after the disk is edited but before the engine completes the operation:
![screenshot-localhost-3000-2018 10 31-15-10-57](https://user-images.githubusercontent.com/3985964/47812651-58e61b00-dd1f-11e8-9dbe-98b1a73754eb.png)

Once the disk is ready to use:
![screenshot-localhost-3000-2018 10 31-15-19-45](https://user-images.githubusercontent.com/3985964/47813096-69e35c00-dd20-11e8-8efc-048ad58b4bb7.png)

Confirm on delete of disk:
![screenshot-localhost-3000-2018 10 31-15-20-19](https://user-images.githubusercontent.com/3985964/47813823-7a94d180-dd22-11e8-916b-cea0e95f3e26.png)

After confirming delete and the delete command is given and before the disk is removed:
![screenshot-localhost-3000-2018 10 31-15-36-20](https://user-images.githubusercontent.com/3985964/47813923-c34c8a80-dd22-11e8-87f2-b8c29d6565e5.png)

sjd78 commented:
(squashed history and rebased, I had too many small commits after addressing code review concerns and finding a few minor issues that needed to be fixed)

sjd78 commented:
ci build please

lwrigh commented:
@sjd78 This is looking great! I just have a few comments.  

For the 'Create New Disk' modal: 

- The field inputs look good to me but I would include info icons next to the storage domain and disk type fields and give the user brief explanations that they can't be edited once they have been selected. The text for the storage domain info tip could read something along the lines of 'Once you have selected a storage domain, you will not be able to edit it.' The same sort of message could be applied to the disk type. 

- I would also include an info tip next to the 'Size' field and include a message telling the user that they can only extend the size of the disk once it has been created. It could read along the lines of 'Once you have created a disk, you can only extend the size of the disk if you make edits to it.'

For the 'Edit Disk' modal: 

- I would get rid of the 'Extend size by' field and only have the 'Size' field. The 'Size' field could feature the same info tip that I noted in the comment above about 'Once you have created a disk, you can only extend the size of the disk if you make any edits to it.' The 'Size' field in edit mode could feature the up and down value adding buttons but have the down arrow disabled so the user could only increase the value. An error message could also appear below the field if they enter a value that is below the current size. 

One tiny thing on the confirm modal, the title should be 'Confirm Delete' so both words are capitalized.

The 'Edit Disk' modal for a DirectLUN disk and the disk states in the card look good to me.

sjd78 commented:
> @sjd78 This is looking great! I just have a few comments.
> 
> For the 'Create New Disk' modal:
> 
> * The field inputs look good to me but I would include info icons next to the storage domain and disk type fields and give the user brief explanations that they can't be edited once they have been selected. The text for the storage domain info tip could read something along the lines of 'Once you have selected a storage domain, you will not be able to edit it.' The same sort of message could be applied to the disk type.
> 
> * I would also include an info tip next to the 'Size' field and include a message telling the user that they can only extend the size of the disk once it has been created. It could read along the lines of 'Once you have created a disk, you can only extend the size of the disk if you make edits to it.'
> 

Added field help in create mode for all 3 fields:
![ezgif com-video-to-gif 1](https://user-images.githubusercontent.com/3985964/47883188-88218880-de02-11e8-9f3f-3d8aa750ad2a.gif)


> 
> For the 'Edit Disk' modal:
> 
> * I would get rid of the 'Extend size by' field and only have the 'Size' field. The 'Size' field could feature the same info tip that I noted in the comment above about 'Once you have created a disk, you can only extend the size of the disk if you make any edits to it.' The 'Size' field in edit mode could feature the up and down value adding buttons but have the down arrow disabled so the user could only increase the value. An error message could also appear below the field if they enter a value that is below the current size.
> 
> 
> One tiny thing on the confirm modal, the title should be 'Confirm Delete' so both words are capitalized.

Done!
![screenshot-localhost-3000-2018 11 01-18-19-02](https://user-images.githubusercontent.com/3985964/47883218-a4bdc080-de02-11e8-99f6-322ea686b6a7.png)

> 
> The 'Edit Disk' modal for a DirectLUN disk and the disk states in the card look good to me.

:+1:

bond95 commented on src/components/VmDetails/cards/DisksCard/DiskImageEditor.js:
```diff
@@ -0,0 +1,428 @@
+              </LabelCol>
+              <Col sm={9}>
+                { (createMode || isImage) &&
+                  <FormControl
```
I still not see here id.

sjd78 replied:
And you won't, but it is there:
![screenshot from 2018-11-02 08-59-05](https://user-images.githubusercontent.com/3985964/47916787-be501e00-de7d-11e8-8a04-5f631b990922.png)

When you use a `FormGroup` with a `FormControl`, the id is specified on the `FormGroup` instead: https://react-bootstrap.github.io/components/forms/#forms-props-form-group

bond95 commented on src/components/VmDetails/cards/DisksCard/DiskImageEditor.js:
```diff
@@ -0,0 +1,428 @@
+                  />
+                }
+                { !createMode && !isImage &&
+                  <div id={`${idPrefix}-size-readonly`} className={style['editor-field-read-only']}>
```
For consistency, please, use `${idPrefix}-size` for readonly values, and `${idPrefix}-size-edit` for form fields.

sjd78 commented:
ci build please

sjd78 commented:
RPMs:
 - [ovirt-web-ui-1.5.0-0.20181102gitf20b911.el7.noarch.rpm](https://jenkins.ovirt.org/job/oVirt_ovirt-web-ui_standard-check-pr/166/artifact/build-artifacts.el7.x86_64/ovirt-web-ui-1.5.0-0.20181102gitf20b911.el7.noarch.rpm)
 - [ovirt-web-ui-1.5.0-0.20181102gitf20b911.fc28.noarch.rpm](https://jenkins.ovirt.org/job/oVirt_ovirt-web-ui_standard-check-pr/166/artifact/build-artifacts.fc28.x86_64/ovirt-web-ui-1.5.0-0.20181102gitf20b911.fc28.noarch.rpm)

bond95 commented on src/components/VmDetails/cards/DisksCard/DiskImageEditor.js:
```diff
@@ -296,7 +296,7 @@ class DiskImageEditor extends Component {
                 }
                 { !createMode && !isImage &&
-                  <div id={`${idPrefix}-size-readonly`} className={style['editor-field-read-only']}>
+                  <div id={`${idPrefix}-size-edit`} className={style['editor-field-read-only']}>
```
Are you sure this is `edit`?

sjd78 replied:
Humm. I don't think that block is even needed anymore.  I'll remove it and leave just the `FormControl` in place.

bond95 commented on src/components/VmDetails/cards/DisksCard/DiskImageEditor.js:
```diff
@@ -0,0 +1,428 @@
+                  />
+                }
+                { !createMode && !isDirectLUN &&
+                  <div id={`${idPrefix}-storage-domain-readonly`} className={style['editor-field-read-only']}>
```
Remove here `readonly` too, please.

bond95 commented on src/components/VmDetails/cards/DisksCard/DiskImageEditor.js:
```diff
@@ -0,0 +1,428 @@
+                  </div>
+                }
+                { isDirectLUN &&
+                  <div id={`${idPrefix}-storage-domain-readonly`} className={style['editor-field-read-only']}>
```
Likewise

sjd78 commented:
ci build please

sjd78 commented:
ci build please

sjd78 commented:
RPMs:
 - [ovirt-web-ui-1.5.0-0.20181102gitf5b48c0.el7.noarch.rpm](https://jenkins.ovirt.org/job/oVirt_ovirt-web-ui_standard-check-pr/172/artifact/build-artifacts.el7.x86_64/ovirt-web-ui-1.5.0-0.20181102gitf5b48c0.el7.noarch.rpm)
 - [ovirt-web-ui-1.5.0-0.20181102gitf5b48c0.fc28.noarch.rpm](https://jenkins.ovirt.org/job/oVirt_ovirt-web-ui_standard-check-pr/172/artifact/build-artifacts.fc28.x86_64/ovirt-web-ui-1.5.0-0.20181102gitf5b48c0.fc28.noarch.rpm)

sjd78 commented:
(rebased to get the new rpm build changes to use the PR's last commit hash in the file name instead of the merge commit hash)

sjd78 commented:
ci build please

sjd78 commented:
RPMs
 - [ovirt-web-ui-1.5.0-0.20181112git80ab0d2.el7.noarch.rpm](https://jenkins.ovirt.org/job/oVirt_ovirt-web-ui_standard-check-pr/202/artifact/build-artifacts.el7.x86_64/ovirt-web-ui-1.5.0-0.20181112git80ab0d2.el7.noarch.rpm)
 - [ovirt-web-ui-1.5.0-0.20181112git80ab0d2.fc28.noarch.rpm](https://jenkins.ovirt.org/job/oVirt_ovirt-web-ui_standard-check-pr/202/artifact/build-artifacts.fc28.x86_64/ovirt-web-ui-1.5.0-0.20181112git80ab0d2.fc28.noarch.rpm)

sjd78 commented:
@smacko - The current rpm build, liked to above, fixes #851 and fixes #852.
